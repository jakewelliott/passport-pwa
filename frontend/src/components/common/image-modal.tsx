import React from 'react';
import { TransformWrapper, TransformComponent } from "react-zoom-pan-pinch";

interface ImageModalProps {
  photo: {
    url: string;
    caption?: string;
  };
  onClose: () => void;
}

export const ImageModal: React.FC<ImageModalProps> = ({ photo, onClose }) => {
  return (
    <div className="fixed inset-0 bg-system_black flex items-center justify-center" style={{zIndex: 9999}}>
      <span className="absolute top-4 right-6 text-system_white text-h1 font-bold cursor-pointer z-10" onClick={onClose}>&times;</span>
      <TransformWrapper>
        <TransformComponent>
          <img 
            src={photo.url} 
            alt={photo.caption || 'Photo'} 
            className="max-w-full max-h-screen object-contain"
          />
        </TransformComponent>
      </TransformWrapper>
      {photo.caption && (
        <div className="absolute bottom-4 left-0 right-0 text-center text-system_white bg-system_black bg-opacity-50 p-2">
          {photo.caption}
        </div>
      )}
    </div>
  );
};
