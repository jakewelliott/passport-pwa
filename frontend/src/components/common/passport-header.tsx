export const PassportHeader = () => {
    return (
        <div className='flex flex-col items-center justify-center h-2/4'>
            <img src="./DPRLogoWhite.svg" alt="North Carolina Department of Parks and Rec White Logo" width="136" height="103" />
            <h3 className="text-center">North Carolina State Parks</h3>
            <p className='script h-20' style={{ paddingTop: 30 }}>Passport</p>
        </div>
    );
}
