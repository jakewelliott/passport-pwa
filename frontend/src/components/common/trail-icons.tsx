import type { Trail } from '@/lib/mock/types';

const ICON_SVG = [
  'Biking-Red',
  'BoatRamp-Blue',
  'BoatRental-Blue',
  'CamperCabins-Green',
  'Camping-Black',
  'Camping-Green',
  'CanoeinCamping-Green',
  'EquestrianCamping-Green',
  'Exhibits-Blue',
  'FBST-Blaze',
  'FFST-Blaze',
  'Fishing-Red',
  'GroupCabins-Green',
  'GroupCamp-Green',
  'HGST-Blaze',
  'Hiking-Red',
  'HorsebackRiding-Red',
  'MST-Blaze',
  'Paddling-Red',
  'PicnicShelter-Black',
  'Picnicking-Red',
  'Playground-Blue',
  'PrimitiveCabin-Green',
  'RockClimbing-Red',
  'RVCamping-Green',
  'Swimming-Red',
  'VacationCabin-Green',
  'VisitorCenter-Blue',
  'YST-Blaze',
] as const;

type IconName = (typeof ICON_SVG)[number];

const filterIcons = (r: string) => ICON_SVG.filter((icon): icon is IconName => icon.includes(r));

const activityIcons = filterIcons('Red');
const amenityIcons = filterIcons('Blue');
const campingIcons = filterIcons('Green');
const blazeIcons = filterIcons('Blaze');

type IconSize = 'sm' | 'md' | 'lg';

const sizeMap = {
  sm: {
    icon: 32,
    text: 'text-xs',
  },
  md: {
    icon: 48,
    text: 'text-sm',
  },
  lg: {
    icon: 64,
    text: 'text-base',
  },
} as const;

const parseText = (fname: string) => {
  // Remove color suffixes and hyphens
  const withoutColor = fname.replace(/-(Red|Blue|Green|Black|Blaze)/g, '');
  // Add spaces before capital letters and remove remaining hyphens
  return withoutColor
    .replace(/([A-Z])/g, ' $1') // Add space before capitals
    .replace(/-/g, ' ') // Remove any remaining hyphens
    .trim(); // Remove leading/trailing spaces
};

export const TrailIcon = ({
  iconName,
  size = 'md',
  showText = false,
}: {
  iconName: IconName;
  size?: IconSize;
  showText?: boolean;
}) => {
  return (
    <div className='flex flex-col items-center gap-1'>
      <div style={{ height: sizeMap[size].icon, width: sizeMap[size].icon }} className='aspect-square'>
        <img src={`/icons/park/${iconName}.svg`} alt={parseText(iconName)} />
      </div>
      {showText && <div className={`${sizeMap[size].text} text-center`}>{parseText(iconName)}</div>}
    </div>
  );
};

export const TrailIcons = ({
  trail,
  className = '',
  size = 'md',
  showText = true,
}: {
  trail: Trail;
  className?: string;
  size?: IconSize;
  showText?: boolean;
}) => {
  if (!trail.trailIcons?.length) return null;

  return (
    <div className={`flex items-center gap-2 ${className}`}>
      {trail.trailIcons.map((iconName) => (
        <TrailIcon key={iconName} iconName={iconName as IconName} size={size} showText={showText} />
      ))}
    </div>
  );
};
