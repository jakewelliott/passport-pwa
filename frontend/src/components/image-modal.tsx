import { a11yOnClick } from '@/lib/a11y';
import type React from 'react';
import { TransformComponent, TransformWrapper } from 'react-zoom-pan-pinch';

interface ImageModalProps {
    photo: {
        photoPath: string;
        alt?: string;
    };
    onClose: () => void;
}

export const ImageModal: React.FC<ImageModalProps> = ({ photo, onClose }) => {
    return (
        <dialog
            aria-modal='true'
            className='fixed inset-0 flex items-center justify-center bg-system_black'
            style={{ zIndex: 9999 }}
        >
            <span
                className='absolute top-4 right-6 z-10 cursor-pointer font-bold text-h1 text-system_white'
                style={{ textShadow: '0 0 4px #444444' }}
                {...a11yOnClick(onClose)}
            >
                &times;
            </span>
            <TransformWrapper>
                <TransformComponent>
                    <img
                        src={photo.photoPath}
                        alt={photo.alt || 'Photo'}
                        className='max-h-screen max-w-full object-contain'
                    />
                </TransformComponent>
            </TransformWrapper>
            {photo.alt && (
                <div
                    data-testid='caption-container'
                    className='absolute right-0 bottom-4 left-0 bg-system_black bg-opacity-50 p-2 text-center text-system_white'
                >
                    {photo.alt}
                </div>
            )}
        </dialog>
    );
};
