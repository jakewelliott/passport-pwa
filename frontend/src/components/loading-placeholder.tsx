// TODO: make this the size of the screen and show a spinner
// and then use it instead of null to render a loading state
export const LoadingPlaceholder = ({ what }: { what?: string }) => {
  return (
    <div className='animate-pulse' data-testid='loading-placeholder'>
      <div className='mb-4 h-32 rounded-md bg-gray-200' />
      <div className='mb-4 h-24 rounded-md bg-gray-200' />
      <div className='h-48 rounded-md bg-gray-200' />
      {what && <p className='text-center text-gray-500 text-sm'>Loading {what}...</p>}
    </div>
  );
};
