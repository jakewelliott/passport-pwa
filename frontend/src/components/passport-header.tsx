export const PassportHeader = () => {
    return (
        <div className='flex h-2/4 flex-col items-center justify-center'>
            <img
                src='/DPRLogoWhite.svg'
                alt='North Carolina Department of Parks and Rec White Logo'
                width='136'
                height='103'
            />
            <h3 className='text-center'>North Carolina State Parks</h3>
            <p className='font-Norican text-8xl' style={{ paddingTop: 30 }}>
                Passport
            </p>
        </div>
    );
};
