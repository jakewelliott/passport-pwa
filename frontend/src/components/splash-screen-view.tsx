import { dbg } from '@/lib/debug';

export const SplashScreenView = ({ children }: { children?: React.ReactNode }) => {
    dbg('RENDER', '<SplashScreenView />');

    return (
        <div
            className='fixed inset-0 flex flex-col items-center bg-no-repeat'
            style={{
                backgroundImage: "url('/photos/CRMO_FrontCover.jpg')",
                backgroundSize: '150%',
                backgroundPosition: 'center 25%',
                zIndex: 9999,
            }}
        >
            <div className='flex h-2/4 flex-col items-center justify-center'>
                <img
                    src='/DPRLogoWhite.svg'
                    alt='North Carolina Department of Parks and Rec White Logo'
                    width='136'
                    height='103'
                />
                <h3>North Carolina State Parks</h3>
                <p className='script' style={{ paddingTop: 30 }}>
                    Passport
                </p>
                <div className='mt-20 rounded-xl bg-secondary-lightblue bg-opacity-70 p-3'>{children}</div>
            </div>
        </div>
    );
};
