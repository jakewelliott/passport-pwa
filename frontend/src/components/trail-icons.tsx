import { dbg } from '@/lib/debug';
import type { Trail } from '@/types';

type IconSize = 'sm' | 'md' | 'lg' | 'xs';

export const sizeMap = {
    xs: {
        icon: 16,
        text: 'text-xs',
    },
    sm: {
        icon: 32,
        text: 'text-xs',
    },
    md: {
        icon: 48,
        text: 'text-sm',
    },
    lg: {
        icon: 64,
        text: 'text-base',
    },
} as const;

export const TrailIcon = ({
    iconName,
    size = 'md',
    showText = false,
}: {
    iconName: string;
    size?: IconSize;
    showText?: boolean;
}) => {
    dbg('RENDER', `trail-icon - ${iconName}`);
    return (
        <div className='flex flex-col items-center gap-1'>
            <div style={{ height: sizeMap[size].icon, width: sizeMap[size].icon }} className='aspect-square'>
                <img src={`/icons/misc/${iconName}.svg`} alt={iconName} />
            </div>
            {showText && <div className={`${sizeMap[size].text} text-center`}>{iconName}</div>}
        </div>
    );
};

export const TrailIcons = ({
    trail,
    className = '',
    size = 'md',
    showText = false,
}: {
    trail: Trail;
    className?: string;
    size?: IconSize;
    showText?: boolean;
}) => {
    const iconNames = trail.icons?.map((icon) => icon.toString()) || [];
    dbg('RENDER', `TRAIL-ICONS - ${iconNames.length}`, iconNames);
    if (!iconNames.length) return null;

    return (
        <div className={`inline-flex items-center gap-1 px-1 ${className}`}>
            {trail.icons.map((trailIcon) => (
                <TrailIcon key={trailIcon} iconName={trailIcon} size={size} showText={showText} />
            ))}
        </div>
    );
};
