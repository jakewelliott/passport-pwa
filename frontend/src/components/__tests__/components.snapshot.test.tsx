import { trails } from '@/lib/testing/mock/tables';
import { renderWithClient } from '@/lib/testing/test-wrapper';
import { screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import Header from '../header';
import { ImageModal } from '../image-modal';
import ListRow from '../list-row';
import { LoadingPlaceholder } from '../loading-placeholder';
import { PassportHeader } from '../passport-header';
import RoundedButton from '../rounded-button';
import { SplashScreenView } from '../splash-screen-view';
import TabBar from '../tab-bar';
import { TrailIcons } from '../trail-icons';

const mockTrail = trails[0];

// Mock ResizeObserver
global.ResizeObserver = vi.fn().mockImplementation(() => ({
    observe: vi.fn(),
    unobserve: vi.fn(),
    disconnect: vi.fn(),
}));

// // Mock the useUser hook
// vi.mock('@/hooks/queries/useUser', () => ({
// 	vi.mock('@/hooks/queries/useUser', () => ({
// 		useUser: () => ({
// 			data: { role: 'visitor' },
// 			isLoading: false,
// 		}),
// 		isLoading: false,
// 	}),
// }))

// // Mock the usePageTitle hook
// vi.mock('@/hooks/usePageTitle', () => ({
// 	vi.mock('@/hooks/usePageTitle', () => ({
// 		usePageTitle: () => ({
// 			pageTitle: 'Test Title',
// 			showBackButton: true,
// 		}),
// 		showBackButton: true,
// 	}),
// }))

describe('Component Snapshots', () => {
    it('TabBar renders correctly', () => {
        renderWithClient(<TabBar />);
        expect(screen.getByText('Locations')).toBeInTheDocument();
    });

    it('RoundedButton renders correctly', () => {
        renderWithClient(<RoundedButton title='Test Button' color='secondary_darkteal' />);
        expect(screen.getByText('Test Button')).toBeInTheDocument();
    });

    it('LoadingPlaceholder renders correctly', () => {
        renderWithClient(<LoadingPlaceholder what='test' />);
        expect(screen.getByText('Loading test...')).toBeInTheDocument();
    });

    it('SplashScreen renders correctly', () => {
        renderWithClient(<SplashScreenView />);
        expect(screen.getByText('Loading...')).toBeInTheDocument();
    });

    it('ImageModal renders correctly', () => {
        renderWithClient(<ImageModal photo={{ photo: 'test.jpg', alt: 'Test Image' }} onClose={() => {}} />);
        expect(screen.getByText('Test Image')).toBeInTheDocument();
    });

    it('Header renders correctly', () => {
        renderWithClient(<Header />);
        expect(screen.getByText('Locations')).toBeInTheDocument();
    });

    it('TrailIcons renders correctly', () => {
        renderWithClient(<TrailIcons trail={mockTrail} size='md' showText={true} />);
        expect(screen.getByText('Trail')).toBeInTheDocument();
    });

    it('ListRow renders correctly', () => {
        renderWithClient(
            <ListRow>
                <div>Test List Row</div>
            </ListRow>,
        );
        expect(screen.getByText('Test List Row')).toBeInTheDocument();
    });

    it('PassportHeader renders correctly', () => {
        renderWithClient(<PassportHeader />);
        expect(screen.getByText('Locations')).toBeInTheDocument();
    });
});
