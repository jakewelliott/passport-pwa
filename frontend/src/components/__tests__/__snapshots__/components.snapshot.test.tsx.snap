// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component Snapshots > Header renders correctly 1`] = `
<div>
  <div
    class="fixed top-0 right-0 left-0 z-50"
  >
    <header
      class="relative flex items-center justify-center bg-secondary_darkteal p-4"
      style="height: 50px;"
    >
      <div
        class="absolute left-4"
      >
        <button
          class="flex items-center text-system_white"
          type="button"
        >
          <svg
            class="mr-1"
            data-testid="fa-chevron-left"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 320 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l192 192c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L77.3 256 246.6 86.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-192 192z"
            />
          </svg>
          <span>
            Back
          </span>
        </button>
      </div>
      <h4
        class="text-system_white"
      >
        Test Title
      </h4>
      <div
        class="absolute right-4 w-[70px]"
        data-testid="balance-placeholder"
      >
        <div
          class="relative"
        />
      </div>
    </header>
  </div>
</div>
`;

exports[`Component Snapshots > ImageModal renders correctly 1`] = `
<div>
  <div
    class="fixed inset-0 flex items-center justify-center bg-system_black"
    style="z-index: 9999;"
  >
    <span
      class="absolute top-4 right-6 z-10 cursor-pointer font-bold text-h1 text-system_white"
    >
      Ã—
    </span>
    <div
      class="react-transform-wrapper transform-component-module_wrapper__SPB86 "
    >
      <div
        class="react-transform-component transform-component-module_content__FBWxo "
        style="transform: translate(0px, 0px) scale(1);"
      >
        <img
          alt="Test Image"
          class="max-h-screen max-w-full object-contain"
          src="/photos/test.jpg"
        />
      </div>
    </div>
    <div
      class="absolute right-0 bottom-4 left-0 bg-system_black bg-opacity-50 p-2 text-center text-system_white"
      data-testid="caption-container"
    >
      Test Image
    </div>
  </div>
</div>
`;

exports[`Component Snapshots > ListRow renders correctly 1`] = `
<div>
  <div
    class="flex items-center bg-supporting_lightblue p-3 shadow-md"
    style="border-radius: 15px;"
  >
    <div>
      Test List Row
    </div>
  </div>
</div>
`;

exports[`Component Snapshots > LoadingPlaceholder renders correctly 1`] = `
<div>
  <div
    class="animate-pulse"
    data-testid="loading-placeholder"
  >
    <div
      class="mb-4 h-32 rounded-md bg-gray-200"
    />
    <div
      class="mb-4 h-24 rounded-md bg-gray-200"
    />
    <div
      class="h-48 rounded-md bg-gray-200"
    />
    <p
      class="text-center text-gray-500 text-sm"
    >
      Loading 
      test
      ...
    </p>
  </div>
</div>
`;

exports[`Component Snapshots > PassportHeader renders correctly 1`] = `
<div>
  <div
    class="flex h-2/4 flex-col items-center justify-center"
  >
    <img
      alt="North Carolina Department of Parks and Rec White Logo"
      height="103"
      src="/DPRLogoWhite.svg"
      width="136"
    />
    <h3
      class="text-center"
    >
      North Carolina State Parks
    </h3>
    <p
      class="script h-20"
      style="padding-top: 30px;"
    >
      Passport
    </p>
  </div>
</div>
`;

exports[`Component Snapshots > RoundedButton renders correctly 1`] = `
<div>
  <button
    class="bg-secondary_darkteal flex cursor-pointer select-none items-center justify-center text-center min-w-32 p-4 rounded-full"
    type="button"
  >
    <p
      class="button-text text-system_white"
    >
      Test Button
    </p>
  </button>
</div>
`;

exports[`Component Snapshots > SplashScreen renders correctly 1`] = `
<div>
  <div
    class="fixed inset-0 flex flex-col items-center bg-no-repeat"
    style="background-image: url(/photos/CRMO_FrontCover.jpg); background-size: 150%; z-index: 9999;"
  >
    <div
      class="flex h-2/4 flex-col items-center justify-center"
    >
      <img
        alt="North Carolina Department of Parks and Rec White Logo"
        height="103"
        src="/DPRLogoWhite.svg"
        width="136"
      />
      <h3>
        North Carolina State Parks
      </h3>
      <p
        class="script"
        style="padding-top: 30px;"
      >
        Passport
      </p>
    </div>
  </div>
</div>
`;

exports[`Component Snapshots > TabBar renders correctly 1`] = `
<div>
  <nav
    class="fixed right-0 bottom-0 left-0 bg-secondary_darkteal"
  >
    <ul
      class="flex h-16 items-center justify-around"
    >
      <li>
        <a
          data-discover="true"
          href="/locations"
          style="text-decoration: none;"
        >
          <div
            class="flex flex-col items-center p-2 text-supporting_inactiveblue"
          >
            <svg
              fill="none"
              height="24px"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M3 7l6 -3l6 3l6 -3v13l-6 3l-6 -3l-6 3v-13"
              />
              <path
                d="M9 4v13"
              />
              <path
                d="M15 7v13"
              />
            </svg>
            <span
              class="p-mini"
            >
              Locations
            </span>
          </div>
        </a>
      </li>
      <li>
        <a
          data-discover="true"
          href="/stamps"
          style="text-decoration: none;"
        >
          <div
            class="flex flex-col items-center p-2 text-supporting_inactiveblue"
          >
            <svg
              fill="currentColor"
              height="24px"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="24px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M32 512h448v-64H32v64zm384-256h-66.56c-16.26 0-29.44-13.18-29.44-29.44v-9.46c0-27.37 8.88-53.41 21.46-77.72 9.11-17.61 12.9-38.39 9.05-60.42-6.77-38.78-38.47-70.7-77.26-77.45C212.62-9.04 160 37.33 160 96c0 14.16 3.12 27.54 8.69 39.58C182.02 164.43 192 194.7 192 226.49v.07c0 16.26-13.18 29.44-29.44 29.44H96c-53.02 0-96 42.98-96 96v32c0 17.67 14.33 32 32 32h448c17.67 0 32-14.33 32-32v-32c0-53.02-42.98-96-96-96z"
              />
            </svg>
            <span
              class="p-mini"
            >
              Stamps
            </span>
          </div>
        </a>
      </li>
      <li>
        <a
          data-discover="true"
          href="/more"
          style="text-decoration: none;"
        >
          <div
            class="flex flex-col items-center p-2 text-supporting_inactiveblue"
          >
            <svg
              fill="currentColor"
              height="24px"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="24px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
            <span
              class="p-mini"
            >
              More
            </span>
          </div>
        </a>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`Component Snapshots > TrailIcons renders correctly 1`] = `
<div>
  <div
    class="flex items-center gap-2 "
  >
    <div
      class="flex flex-col items-center gap-1"
    >
      <div
        class="aspect-square"
        style="height: 48px; width: 48px;"
      >
        <img
          alt="Hiking"
          src="/icons/park/Hiking-Red.svg"
        />
      </div>
      <div
        class="text-sm text-center"
      >
        Hiking
      </div>
    </div>
    <div
      class="flex flex-col items-center gap-1"
    >
      <div
        class="aspect-square"
        style="height: 48px; width: 48px;"
      >
        <img
          alt="Camping"
          src="/icons/park/Camping-Green.svg"
        />
      </div>
      <div
        class="text-sm text-center"
      >
        Camping
      </div>
    </div>
  </div>
</div>
`;
